stan/math/prim/mat/prob/multi_normal_rng.hpp

```diff
@@ -33,14 +46,21 @@ namespace stan {
       check_symmetric(function, "Covariance matrix", S);
       check_finite(function, "Location parameter", mu);
 
+      Eigen::LLT<Eigen::MatrixXd> llt_of_S = S.llt();
+      if (llt_of_S.info() != Eigen::Success) {
+        throw std::domain_error("multi_normal_rng:"
```

Comment:
Thanks for the tip.  I'm going too fast and didn't see the "llt_of".
That's a much better way to go.

Down the rabbit hole I go.

Why do these things all return bool values that are never used?  None
of them can return false.  Should I start another issue for that or
am I missing something?

I'm reluctant to change the error message given I don't know what
other functions call this function.  

Isn't the second test here both expensive (probably not compared to
Cholesky factoring, but still) and redundant?  

     if (cholesky.info() != Eigen::Success
          || !(cholesky.matrixLLT().diagonal().array() > 0.0).all())

Wikipedia says you can only Cholesky factor a positive-definite matrix.

And the first test uses "cholesky.isPositive()" and doesn't have the
diagonal().array() > 0 business.  Instead, it has:

          || (cholesky.vectorD().array() <= 0.0).any())

which may not work if any of the results are NaN, because they'll return
false.  It then goes on to do check_not_nan() on the whole function, but
the form used with the diagonal().array() will already check that, because
any NaN will automatically return false, so the test will fail.

If the isPositive() check is different than the diagonal().array() > 0 check,
can we return a different and more relevant error message? 

Also not sure how that passed lint because it had multiple lines longer than
80 chars.  

I'm also confused about the intended usage pattern of domain_error(); in
the check_pos_definite file I found:

        domain_error(function, "LDLT decomposition of", " failed", name);
        domain_error(function, name, "is not positive definite.", "");

and added

        domain_error(function, "Matrix", " is not positive definite", name);

Also, I don't like having to pad one string and not the other---just seems like
it's ripe for errors.

Also, do I need to keep putting "fixes #xyz" in commits or does it just
need to go in the pull request or what? 

- Bob



On Sep 2, 2016, at 6:31 PM, Daniel Lee <notifications@github.com> wrote:
> 
> In stan/math/prim/mat/prob/multi_normal_rng.hpp:
> 
> > @@ -33,14 +46,21 @@ namespace stan {
> >        check_symmetric(function, "Covariance matrix", S);
> >        check_finite(function, "Location parameter", mu);
> >  
> > +      Eigen::LLT<Eigen::MatrixXd> llt_of_S = S.llt();
> > +      if (llt_of_S.info() != Eigen::Success) {
> > +        throw std::domain_error("multi_normal_rng:"
> 
> Want to just use check_pos_definite()? If so, you can replace lines 50-54 with:
> 
>   check_pos_definite("multi_normal_rng", "matrix argument", llt_of_S);
> 
> If you're going down path, could you change the error message on check_pos_definite? What you've written is so much clearer than "Cholesky decomposition of <> failed".
> 
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub, or mute the thread.
>